OCPCLOUD-1674: Update installation docs to account for generation/activation (#123)
openshift/cluster-control-plane-machine-set-operator: docs/user/installation.md

# Installation

> This document provides instructions for users to configure a `ControlPlaneMachineSet` within an OpenShift cluster.

Reviewer: Drop this change please, we spell it out so that it's consistent with the product docs and IBM style guidelines

This process can be followed on OpenShift clusters that are version 4.12 or higher.

A `ControlPlaneMachineSet` can be installed into any existing OpenShift cluster provided it has existing, and `Running`

> control plane machines.

Reviewer: Any? Is this true? What about a GCP cluster for example where we know there are issues. I would probably not say any

Typically, this would only be true if the cluster was created using Installer-Provisioned infrastructure.

> > Note: A `Running` Control Plane Machines means the Machine phase is `Running`. By requiring at least 1 `Running`
> > Machine we can ensure that the spec of the Machine is valid and that the CPMSO will be able to create new Machines
> > based on that template.

Reviewer: ```suggestion
> Note: A `Running` control plane machine means the machine is in the `Running` phase. By requiring at least 1 `Running`
> machine we can ensure that the spec of the machine is valid and that the control plane machine set will be able to create new machines
> based on that template.
```

## Installer based installation

> For newly installed clusters with OpenShift version 4.12 onwards, the installer provisioned infrastructure (IPI) installer workflow will create a
> ControlPlaneMachineSet resource (if the platform is supported) and set it to `Active`.
> So nothing is required from the user side.

Reviewer: ```suggestion
For clusters installed with OpenShift version 4.12 onwards, the installer provisioned infrastructure (IPI) installer workflow will create a
control plane machine set (if the platform is supported) and set it to `Active`.
No action is required by the user for these clusters.
```

> You can issue a get request to check if that's the case:

Reviewer: Issues a get request isn't user speak IMO, I doubt we'd say that in product docs

Perhaps instead we can say

```suggestion
You can check the control plane machine set using the following command:
```

```
oc -n openshift-machine-api get controlplanemachineset cluster

> ```

Author: We could convert these to notes maybe?

```
oc -n openshift-machine-api get controlplanemachineset cluster
```

As of 4.12, the only supported platform is Amazon Web Services (AWS). Support for other platforms will be added in
future releases.

## Installation into an existing cluster

> ### Creating a control plane machine set for an existing cluster

Reviewer: We are dropping quite a bit here, I wonder if this is still helpful for clusters that don't have failure domains or clusters that don't have CPMS generation enabled?

Author: AFAIK the only supported platform configuration is AWS, so I think I'd like to have it mirrors that.
We'll need to extend the doc to include the upcoming platforms under "supported platforms".
I think that will be the right time to refactor this to match more closely what we'll support both in terms of platforms and manual vs. generated CPMSes.

Reviewer: Azure is supported as far as I was concerned, we have docs for manual installation on Azure and it has been manually tested as well. 

I'd like to keep the manual process not only for Azure but for other platforms where the generator and installer work is not yet implemented, yet a user could use the CPMS

> this resource may exist (if the platform is supported)
> but won't be `Active` by default (rather it will be `Inactive`).
> You can issue a get request to check if that's the case:

Reviewer: ```suggestion
the control plane machine set may exist (if the platform is supported)
but will not be `Active` unless it has been activated by a cluster administrator.

You can check the control plane machine set using the following command:
```

```

> oc -n openshift-machine-api get controlplanemachineset cluster

Reviewer: Prefer fully groupified resource names and long form arguments in docs please

```

> If preset you can activate it, but before doing so you must thoroughly review it and check
> that the generated resource spec actually aligns with the desired values.

Reviewer: ```suggestion
If preset, the control plane machine set may be activated.
Before doing so, the control plane machine set spec must be thoroughly reviewed to ensure
that the generated spec aligns with the desired specification.
```

> You can consult [Anatomy of a ControlPlaneMachineSet](#anatomy-of-a-controlplanemachineset) and [Configuring provider specific fields](#configuring-provider-specific-fields) as a reference for understanding fields and values and what to expect for them.

Reviewer: ```suggestion
Consult the [anatomy of a ControlPlaneMachineSet resource](#anatomy-of-a-controlplanemachineset) and [configuring provider specific fields](#configuring-provider-specific-fields) as a reference for understanding the fields and values within a `ControlPlaneMachineSet` resource.
```

> Extra care should be taken to ensure the values match the desired ones.

Reviewer: I think we can skip this, we said it already on line 39 basically

> The generated `ControlPlaneMachineSet` can be reviewed by issuing:

Reviewer: ```suggestion
The generated control plane machine set can be reviewed with the following command:
```

```

> oc -n openshift-machine-api edit controlplanemachineset cluster

Reviewer: Why edit here?

Author: Because later we tell them they should oc edit any value they want to change and then update the `.spec.state`.
I'm seeing this as a unique thing:
- they issue the edit
- the CPMS opens up in their editor and they read the content
- optionally they change what they need
- they update the `.spec.state` to `Active`
- they save and quit the editor and the change gets applied.

Reviewer: Ack ok, might want to have an example of a patch as well, I think a lot of people are uncomfortable with edit, as it isn't really automatable

Author: Offline we decided to leave it as is and monitor the feedback we get from users on the guide.

```

> If any of the fields don't match with the expected ones, you can change them, provided that the edit is done in the
> same `oc edit` session where the `.spec.state` of the `ControlPlaneMachineSet` is also changed to `Active`.

Reviewer: Don't use contractions in docs
```suggestion
If any of the fields do not match with the expected value, the value may be changed, provided that the edit is done in the
same `oc edit` session where the control plane machine set is activated.
```

> Once happy with the result you can activate the `ControlPlaneMachineSet` by
> setting its `.spec.state` field to `Active`, saving and closing.

Reviewer: ```suggestion
Once the spec of the control plane machine set has been reviewed, activate the control plane machine set by setting the `.spec.state` field to `Active`.
```

Once activated, the `ControlPlaneMachineSet` operator should start the reconciliation of the resource.

### Configuring provider specific fields

The following instructions describe how the failure domains and providerSpec fields should be
constructed depending on the platform of the cluster.

You are not required to do any of the following steps as these fields are already automatically constructed
for you and collated in the generated resource.
The procedure on how to construct them is described here, for it to be used as a reference

> while reviewing the generated fields and to instruct you on how they can be modified.

Reviewer: Not always true, we don't have generators on Azure yet so I'd like to make sure we still have appropriate documentation for Azure and other supported but not generated platforms

Author: I was under this impression: https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/123/files#r1004737242

But we can have a chat to discuss this further :)

Before doing so, the control plane machine set spec must be thoroughly reviewed to ensure
that the generated spec aligns with the desired specification.

Consult [Anatomy of a ControlPlaneMachineSet resource](#anatomy-of-a-controlplanemachineset)
as a reference for understanding the fields and values within a `ControlPlaneMachineSet` resource.

The generated control plane machine set can be reviewed with the following command:
```
oc -n openshift-machine-api edit controlplanemachineset cluster
```

If any of the fields do not match with the expected value, the value may be changed, provided that the edit is done in the
same `oc edit` session where the control plane machine set is activated.

Once the spec of the control plane machine set has been reviewed, activate the control plane machine set by setting the `.spec.state` field to `Active`.

Once activated, the `ControlPlaneMachineSet` operator should start the reconciliation of the resource.

### Installation into an existing cluster with manual resource

The control plane machine set may not exist in the cluster (unless a cluster administrator has created one already),

> but it can be manually created and made `Active`.

Reviewer: ```suggestion
but it can be manually created and activated.
```

This can be checked by using the following command:
```
oc get controlplanemachineset.machine.openshift.io cluster --namespace openshift-machine-api
```

> To manually create a control plane machine set define a `ControlPlaneMachineSet` resource as described in [Anatomy of a ControlPlaneMachineSet resource](#anatomy-of-a-controlplanemachineset).

Reviewer: ```suggestion
To manually create a control plane machine set define a `ControlPlaneMachineSet` resource as described in the [anatomy of a ControlPlaneMachineSet resource](#anatomy-of-a-controlplanemachineset).
```

  name: cluster
  namespace: openshift-machine-api
spec:

>   state: Active

Reviewer: Should probably add `[n]` and explain the state field, wdyt?

2. The strategy defaults to `RollingUpdate`. `OnDelete` is also supported.
3. The ControlPlaneMachineSet spreads Machines across multiple failure domains where possible.
This field must be set to the platform name.

> 4. The failure domains vary on their configuration per platform, see [Configuring provider specific field](#configuring-provider-specific-fields) for how to configure a failure domain on each platform.

Reviewer: ```suggestion
4. The failure domains vary on their configuration per platform, see [configuring provider specific fields](#configuring-provider-specific-fields) for how to configure a failure domain on each platform.
```
